modules/rgbd/src/pose_graph.cpp

```diff
@@ -268,9 +303,89 @@ void Optimizer::optimizeGaussNewton(const Optimizer::Params& params, PoseGraph&
             std::cout << "Gauss newton converged \n";
             break;
         }
-        poseGraph = poseGraph.update(delta);
+        poseGraph    = poseGraph.update(delta);
         prevResidual = currentResidual;
+        iter++;
     } while (iter < params.maxNumIters);
 }
+
+void Optimizer::optimizeLevenberg(const Optimizer::Params& params, PoseGraph& poseGraph)
+{
+    PoseGraph poseGraphOriginal = poseGraph;
+
+    if (!poseGraphOriginal.isValid())
+    {
+        CV_Error(Error::StsBadArg, "Invalid PoseGraph that is either not connected or has invalid nodes");
+        return;
+    }
+
+    int numNodes = poseGraph.getNumNodes();
+    int numEdges = poseGraph.getNumEdges();
+
+    std::cout << "Optimizing posegraph with nodes: " << numNodes << " edges: " << numEdges << std::endl;
+    int iter           = 0;
+    float lambda       = 1e-3f;
+    const float factor = 2;
+    float prevResidual = std::numeric_limits<float>::max();
+    do
+    {
+        std::cout << "Current iteration: " << iter << std::endl;
+
+        BlockSparseMat<float, 6, 6> ApproxH(numNodes);
+        Mat B = cv::Mat::zeros(6 * numNodes, 1, CV_32F);
+
+        float currentResidual = poseGraph.createLinearSystem(ApproxH, B);
+        Mat delta(6 * numNodes, 1, CV_32F);
+        Mat X = poseGraph.getVector();
+
+        //! H_LM = H + lambda * diag(H);
+        Mat Hdiag = ApproxH.diagonal();
+        for (int i = 0; i < Hdiag.rows; i++)
+        {
+            ApproxH.refElem(i, i) += lambda * Hdiag.at<float>(i, 0);
+        }
+        bool success = sparseSolve(ApproxH, B, delta);
+        if (!success)
+        {
+            CV_Error(Error::StsNoConv, "Sparse solve failed");
+            return;
+        }
+        std::cout << "\n delta norm: " << cv::norm(delta) << " X norm: " << cv::norm(X)
+                  << std::endl;
+
+        //! If step size is too small break
+        if (isStepSizeSmall(delta, params.minStepSize))
+        {
+            std::cout << "Step size is too small. stopping\n";
+            break;
+        }
+        std::cout << " Current residual: " << currentResidual << " Prev residual: " << prevResidual << "\n";
```

Comment:
Is it OK that `currentResidual` is calculated before the system is solved? I guess this should be done after obtaining new X.